pkg/agent/packetcapture/packetcapture_controller.go

```diff
@@ -0,0 +1,771 @@
+	var file afero.File
+	if _, err := os.Stat(filePath); err == nil {
+		klog.Warningf("The packet file %s already exists. This may be caused by an unexpected termination, will delete it", filePath)
+		if err := defaultFS.Remove(filePath); err != nil {
+			return nil, nil, err
+		}
+	}
+	file, err := defaultFS.Create(filePath)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create pcapng file: %w", err)
+	}
+	writer, err := pcapgo.NewNgWriter(file, layers.LinkTypeEthernet)
+	if err != nil {
+		return nil, nil, fmt.Errorf("couldn't initialize a pcap writer: %w", err)
+	}
+	return file, writer, nil
+}
+
+// getTargetCaptureDevice is trying to locate the target device for packet capture. If the target
+// Pod does not exist on the current Node, the agent on this Node will not perform the capture.
+// In the PacketCapture spec, at least one of `.Spec.Source.Pod` or `.Spec.Destination.Pod`
+// should be set.
+func (c *Controller) getTargetCaptureDevice(pc *crdv1alpha1.PacketCapture) *string {
+	var pod, ns string
+	if pc.Spec.Source.Pod != nil {
+		pod = pc.Spec.Source.Pod.Name
+		ns = pc.Spec.Source.Pod.Namespace
+	} else {
+		pod = pc.Spec.Destination.Pod.Name
+		ns = pc.Spec.Destination.Pod.Namespace
+	}
+
+	podInterfaces := c.interfaceStore.GetContainerInterfacesByPod(pod, ns)
+	if len(podInterfaces) == 0 {
+		return nil
+	}
+
+	return &podInterfaces[0].InterfaceName
+}
+
+func (c *Controller) startPacketCapture(ctx context.Context, pc *crdv1alpha1.PacketCapture, device *string) error {
+	klog.V(4).InfoS("Started processing PacketCapture", "name", pc.Name)
+	pcState := c.captures[pc.Name]
+	pcState.name = pc.Name
+	srcIP, dstIp, err := c.parseIPs(pc)
+	if err != nil {
+		return err
+	}
+	klog.V(2).InfoS("Prepare capture on the current Node", "name", pc.Name, "device", *device)
+	pcState.targetCapturedPacketsNum = pc.Spec.CaptureConfig.FirstN.Number
+	file, writer, err := getPacketFileAndWriter(pc.Name)
+	if err != nil {
+		return err
+	}
+	pcState.pcapngFile = file
+	pcState.pcapngWriter = writer
+	pcState.updateRateLimiter = rate.NewLimiter(rate.Every(captureStatusUpdatePeriod), 1)
+	go func() {
+		captureErr := c.performCapture(ctx, pc, pcState, *device, srcIP, dstIp)
+		func() {
+			c.lock.Lock()
+			defer c.lock.Unlock()
+			c.numRunningCaptures -= 1
+			state := c.captures[pc.Name]
+			state.phase = packetCapturePhaseCompleted
+			state.err = captureErr
+		}()
+		c.addPacketCapture(pc)
+	}()
+	return nil
+}
+
+func (c *Controller) performCapture(
+	ctx context.Context,
+	pc *crdv1alpha1.PacketCapture,
+	captureState *packetCaptureState,
+	device string,
+	srcIP, dstIP net.IP,
+) error {
+	packets, err := c.captureInterface.Capture(ctx, device, srcIP, dstIP, pc.Spec.Packet)
+	if err != nil {
+		klog.ErrorS(err, "Failed to start capture")
+		return err
+	}
+
+	for {
+		select {
+		case packet := <-packets:
+			c.lock.Lock()
+			if captureState.isCaptureSucceed() {
+				c.lock.Unlock()
+				return nil
+			}
```

Comment:
It's contradictory to check if the capture succeeds before counting the new packet. It could have exited the loop when it receives the last packet.

Reply:
removed.